### ISSUE TITLE
[no-path-concat] overlooks template literals

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#12323]
**Tell us about your environment**

* **ESLint Version:** 6.4.0
* **Node Version:** -
* **npm Version:** -

**What parser (default, Babel-ESLint, etc.) are you using?**

Default.

**Please show your full configuration:**
**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcGF0aC1jb25jYXQ6IGVycm9yICovXG5cbmNvbnN0IGEgPSBfX2Rpcm5hbWUgKyBcIi9mb28uanNcIlxuY29uc3QgYiA9IGAke19fZGlybmFtZX0vZm9vLmpzYFxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMSwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

```js
/*eslint no-path-concat: error */

const a = __dirname + "/foo.js"
const b = `${__dirname}/foo.js`
```

**What did you expect to happen?**

There are two errors:

- `__dirname + "/foo.js"`
- `` `${__dirname}/foo.js` ``

**What actually happened? Please include the actual, raw output from ESLint.**

There is only one error:

- `__dirname + "/foo.js"`

The rule doesn't seem to consider the interpolation of template literals as concatenation.

**Are you willing to submit a pull request to fix this bug?**

Yes.

---

Should we pass the rule to `eslint-plugin-node` plugin?

**Comment by g-plane** [comment]
I'm :+1: on transferring this rule to `eslint-plugin-node`.

**Comment by kaicataldo** [comment]
:+1: to updating the rule's behavior and am not opposed to deprecating it in core and moving it to `eslint-plugin-node`. I wonder if it's worth evaluating whether we should move our other Node-specific rules to the plugin?

**Comment by lekterable** [comment]
what would be the usual flow of `deprecating` the rule in the core package @kaicataldo ? I'd love to work on this if I could get some details that would help me get me started

**Comment by gabrieldrs** [comment]
Hey, I was fiddling around with this earlier today and I thought that maybe we would want to fix the rule first, and then decide to move it to the plugin?

Anyway, I tried to draft an initial solution for this and came up with https://github.com/gabrieldrs/eslint/commit/55a3dbf8e0c26a078ab43a0024fbe95e6eddfe3e

I'm still a newbie with ESTree so my code may have flaws that I'm not aware (any direction would be really appreciated).

As a side note
Running `npm test` showed that eslint uses this syntax in a few places:

```
[...]/eslint/tests/bin/eslint.js
   11:49  error  Use path.join() or path.resolve() instead of + to create paths  no-path-concat
  169:47  error  Use path.join() or path.resolve() instead of + to create paths  no-path-concat
  185:30  error  Use path.join() or path.resolve() instead of + to create paths  no-path-concat
  362:35  error  Use path.join() or path.resolve() instead of + to create paths  no-path-concat

[...]/eslint/tests/lib/cli-engine/config-array/config-array.js
  697:39  error  Use path.join() or path.resolve() instead of + to create paths  no-path-concat
  698:39  error  Use path.join() or path.resolve() instead of + to create paths  no-path-concat
  725:39  error  Use path.join() or path.resolve() instead of + to create paths  no-path-concat
```

**Comment by kaicataldo** [comment]
Thanks for working on this. Your implementation generally looks good to me! Would you like to make a formal PR? We'll need to add some tests to cover the new behavior and fix up any instances of the error in our codebase.

**Comment by gabrieldrs** [comment]
ok, I will work on the tests and fixing the error in the code base today and will open a pr with the changes :)

**Comment by kaicataldo** [comment]
Since this issue was accepted, the TSC has decided to deprecate all the Node-specific rules. eslint-plugin-node will be adding these rules, and our official recommendation will be to switch to that plugin. For more information regarding this change, please see https://github.com/eslint/eslint/issues/12835.
